new feature - bookmark shortcut on long press

pastmaster007 opened the pull request:
**Description**
Now app prompts to save as a bookmark when long press detected on bookmarks toolbar icon..\
App displays toast message when a web page already bookmarked..

**Steps to test this PR**:
<https://vimeo.com/318328577>


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

Commit:
new feature - quick bookmark on long press

now app prompts to save as a bookmark when long press detected on bookmarks toolbar icon..
App displays toast message when a web page already bookmarked..
adds below two types of pixel tags for analytics
- mlp_bm_s //MobileLongPress_BookMark_Save
- mt_bm_l //Mobile_Tabbar_BookMark_Longpress

Commit:
removed pixel tracking..
removed .init..
removed unwanted logging..

Commit:
merging find-in-page changes from develop branch..

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+    
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+//        Logger.log(text: "touchesBegan \(Date.init())")
```
Please remove code that is commented out.

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,91 @@
+//
+//  Created by BG on 2/19/19.
+//  Copyright © 2019 DuckDuckGo. All rights reserved.
+//
```
Please update copyright based on what's in other source files.

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,91 @@
+
+    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
+        
+    }
```
Please remove empty functions.

bwaresiak commented on DuckDuckGo/MainViewController.swift:
```diff
@@ -164,6 +167,12 @@ class MainViewController: UIViewController {
+    private func initBookmarksButton() {
+        gestureBookmarksButton.delegate = self
+        gestureBookmarksButton.image = UIImage(named: "Bookmarks")
+        bookmarksButton.customView = gestureBookmarksButton
```
The one problem with current implementation is that Bookmarks button no longer has a "highlighted" state - there is no visual feedback in case it is pressed.

pastmaster007 replied:
added animation using image view alpha on touch..

bwaresiak commented on DuckDuckGo/TabViewControllerLongPressBookmarkExtension.swift:
```diff
@@ -0,0 +1,36 @@
+                saveCompletion: saveCompletion)
+            self.present(alert, animated: true, completion: nil)
+        } else {
+            Logger.log(text: "Invalid bookmark link found on bookmark long press")
```
I don't think we need to log anything in this method - these are quite common scenarios.

pastmaster007 replied:
i think this error condition would be useful if we ever get into such situations, unless there is already some kind of error handling exists when such scenario happens.
As i know, these logging messages are only enabled in debug build, if it occurs.. Let me know your thoughts..

bwaresiak reviewed (changes requested):
Thank you for your PR. Again, I've put the comments inline. :)

Commit:
updated license info on new files..
added toolbar button touch animation..
display error message when no tab found to bookmark..
removed extra logging and empty functions..

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,112 @@
+    }
+    
+    fileprivate func imposeReleaseAnimation() {
+        iconImageView.alpha = 1.0
```
I'd wrap both press and release animations in UIView.animate for a duration of 0.3, to match how fire button is being animated when pressed.

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,112 @@
+
+    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
+        guard let touch = touches.first else { return }
+        guard point(inside: touch.location(in: self), with: event) else { return }
```
I believe we need to call `imposeReleaseAnimation()` in the `else`, otherwise image won't be returned to normal state in case user taps&holds on a button, then swipes his finger away.

bwaresiak commented on DuckDuckGo/MainViewController.swift:
```diff
@@ -273,10 +282,10 @@ class MainViewController: UIViewController {
+//    @IBAction func onTabBarBookmarksPressed() {
+//        Pixel.fire(pixel: .tabBarBookmarksPressed)
+//        onBookmarksPressed()
+//    }
```
Please remove commented out code.

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,112 @@
+
+import UIKit
+
+protocol GestureToolBarButtonDelegate: NSObjectProtocol {
```
I would consider matching UIKit naming of `UIToolbar` (lowercase 'b'): `GestureToolbarButtonDelegate` here and below `GestureToolbarButton`

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,112 @@
+protocol GestureToolBarButtonDelegate: NSObjectProtocol {
+    
+    func singleTapHandler()
+    func longPressHandler()
```
Generally, for delegate design pattern we should add info which object is the sender of a message, example:
`func singleTapDetected(in: GestureToolBarButton)`

This way, in case we are using multiple GestureToolBarButtons, we can differentiate between them in delegate methods.

bwaresiak commented on DuckDuckGo/MainViewController.swift:
```diff
@@ -793,6 +802,24 @@ extension MainViewController: TabSwitcherButtonDelegate {
+    func longPressHandler() {
+        
+        guard currentTab != nil else {
+            self.view.showBottomToast(UserText.webSaveBookmarkNone)
```
self is not needed here

bwaresiak commented on DuckDuckGo/GestureToolBarButton.swift:
```diff
@@ -0,0 +1,112 @@
+        let longPressRecognizer = UILongPressGestureRecognizer(target: self, action: #selector(longPressHandler))
+        longPressRecognizer.minimumPressDuration = Constants.minLongPressDuration
+        longPressRecognizer.allowableMovement = CGFloat(Constants.maxTouchDeviationPoints)
+        self.addGestureRecognizer(longPressRecognizer)
```
self is not needed here

bwaresiak commented on DuckDuckGo/TabViewControllerLongPressBookmarkExtension.swift:
```diff
@@ -0,0 +1,46 @@
+                title: UserText.alertSaveBookmark,
+                bookmark: link,
+                saveCompletion: saveCompletion)
+            self.present(alert, animated: true, completion: nil)
```
self is not needed here (and same in line 29)

bwaresiak reviewed (changes requested):
Thank you for your patience and once again sorry it took me so long to get back to you.

I've re-run the code and gave it another thought - you can see my comments inline. Other than that it works really well, so I believe that this can be merged once everything is resolved.

Thank you! :)

Commit:
Merge branch 'develop' into feature/quickbookmark

* develop:
  New Feature - quick sharing bookmark from bookmarks screen.. (#409)
  removed as unclear exactly what happens
  upload to test flight
  tweak release lane to match removed beta lane
  clean up Fastlane config and move to automatic and phased release
  bump version and update release notes
  improve onboarding experiment 1 (#419)
  fixes for ipad (#415)

Commit:
fixed alpha animation issue on gesture tool bar button during long press detection when touch end event fires on different view..
removed unnecessary self usage..
added timed animation for gesture toolbar press
renamed gesture toolbar button class name and delegate name to match bar in UIToolbar class..
changed gesture tool bar delegate methods signature to pass itself as parameter to delegate..
removed extra foundation import in networks tests file..
removed commented code..

bwaresiak reviewed (approved):
Everything looks great, thank you for your contribution. :)
